content/en/blog/_posts/2024-02-29-cri-o-seccomp-oci-artifacts.md

```diff
@@ -0,0 +1,421 @@
+solve that problem by running as a daemon within the cluster, which makes me
+wonder which part of that distribution could be done by the [container
+runtime](/docs/setup/production-environment/container-runtimes).
+
+Runtimes usually apply the profiles from a local path on-disk, for example:
+
+```yaml
+apiVersion: v1
+kind: Pod
+metadata:
+  name: pod
+spec:
+  containers:
+    - name: container
+      image: nginx:1.25.3
+      securityContext:
+        seccompProfile:
+          type: Localhost
+          localhostProfile: nginx-1.25.3.json
+```
+
+The profile `nginx-1.25.3.json` has to be available in the root directory of the
+kubelet, appended by the `seccomp` directory. Means the default location for the
+profile on-disk would be `/var/lib/kubelet/seccomp/nginx-1.25.3.json`. If the
+profile is not available, then runtimes will fail on container creation like
+this:
+
+```
+> kubectl get pods
+NAME   READY   STATUS                 RESTARTS   AGE
+pod    0/1     CreateContainerError   0          38s
+```
+
+```
+
+> kubectl describe pod/pod | tail
+Tolerations:                 node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
+                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
+Events:
+  Type     Reason     Age                 From               Message
+  ----     ------     ----                ----               -------
+  Normal   Scheduled  117s                default-scheduler  Successfully assigned default/pod to 127.0.0.1
+  Normal   Pulling    117s                kubelet            Pulling image "nginx:1.25.3"
+  Normal   Pulled     111s                kubelet            Successfully pulled image "nginx:1.25.3" in 5.948s (5.948s including waiting)
+  Warning  Failed     7s (x10 over 111s)  kubelet            Error: setup seccomp: unable to load local profile "/var/lib/kubelet/seccomp/nginx-1.25.3.json": open /var/lib/kubelet/seccomp/nginx-1.25.3.json: no such file or directory
+  Normal   Pulled     7s (x9 over 111s)   kubelet            Container image "nginx:1.25.3" already present on machine
+```
+
+The major obstacle of having to manually distribute the profiles will lead many
+end-users into using `RuntimeDefault` or even running their workloads as
+`Unconfined`.
+
+## CRI-O to the rescue
+
+The Kubernetes container runtime [CRI-O](https://github.com/cri-o/cri-o)
+provides various features using custom annotations. The v1.30 release
+adds [support for a new annotation](https://github.com/cri-o/cri-o/pull/7719)
+called `io.kubernetes.cri-o.seccompProfile`. This annotation allows to specify:
+
+- a seccomp profile for a specific container, when used as:
+  `io.kubernetes.cri-o.seccompProfile/<CONTAINER>`
+- a seccomp profile for every container within a pod, when used without the
+  container name suffix:
+  `io.kubernetes.cri-o.seccompProfile`
+- a seccomp profile for a whole container image, if the image itself contains
+  the annotation `io.kubernetes.cri-o.seccompProfile`
```

Comment:
We can still change the annotation because the feature has not been released. Does that mean that something like this would be more feasible:

- `runtime.kubernetes.io/secocmpProfile/<CONTAINER>`
- `seccompProfile.cri-o.kubeneretes.io/<CONTAINER>`

:thinking: 

cc @haircommander

Reply:
I'd go for something like one of:
- `kubernetes.cri-o.io/seccomp-profile-map` (which would be a JSON dict of container names to profile names)
- `kubernetes.cri-o.io/seccomp-profile-FOO` which would be the seccomp profile for a container named `FOO`

These are CRI-O's annotations, AIUI, so the domain would be a CRI-O one. However we might add a more generic `kubernetes.cri-o.io/seccomp-profile-map` annotation and teach the kubelet how to expose that via CRI (future work).

Reply:
I know these are different to existing annotations; all the existing ones are outside what we (Kubernetes) recommend doing.